sklearn/utils/validation.py

```diff
@@ -97,18 +98,29 @@ def _assert_all_finite(
 
     if _get_config()["assume_finite"]:
         return
+
     X = np.asanyarray(X)
     # First try an O(n) time, O(1) space solution for the common case that
-    # everything is finite; fall back to O(n) space np.isfinite to prevent
-    # false positives from overflow in sum method.
+    # everything is finite; fall back to O(n) space `np.isinf/isnan` or custom
+    # Cython implementation to prevent false positives and provide a detailed
+    # error message.
     is_float = X.dtype.kind in "fc"
-    with np.errstate(over="ignore"):
-        first_pass_isfinite = is_float and np.isfinite(np.sum(X))
-    if first_pass_isfinite:
-        pass
-    elif is_float:
-        has_inf = np.isinf(X).any()
-        has_nan = np.isnan(X).any()
+    # Cython implementation doesn't support FP16 or complex numbers
+    first_pass_isfinite = False
+    if is_float:
+        use_cython = X.dtype in {np.float32, np.float64}
```

Comment:
I think the cython assumes contiguous as well? If so:

```suggestion
        use_cython = X.data.contiguous and X.dtype.type in {np.float32, np.float64}
```

Also the way that NumPy types work, the check needs to be: `X.dtype.type in {np.float32, np.float64}`